server/validate/tel/validate.go

```diff
@@ -2,58 +2,207 @@
+		}
+	}
+
+	if v.Sender == "" {
+		v.Sender = defaultSender
+	}
+	if v.MaxRetries == 0 {
+		v.MaxRetries = defaultMaxRetries
+	}
+	if v.CodeLength == 0 {
+		v.CodeLength = defaultCodeLength
+	}
+	v.maxCodeValue = big.NewInt(0).Exp(big.NewInt(10), big.NewInt(int64(v.CodeLength)), nil)
 
 	return nil
 }
 
+// IsInitialized returns true if the validator is initialized.
+func (v *validator) IsInitialized() bool {
+	return v.CodeLength > 0
+}
+
 // PreCheck validates the credential and parameters without sending an SMS or making the call.
-// If credential is valid it's formatted and prefixed with a tag namespace.
+// If credential is valid, it's formatted and prefixed with a tag namespace.
 func (*validator) PreCheck(cred string, params map[string]interface{}) (string, error) {
+	// Parse will try to extract the number from any text, make sure it's just the number.
+	if !phonenumbers.VALID_PHONE_NUMBER_PATTERN.MatchString(cred) {
+		return "", t.ErrMalformed
+	}
 	countryCode, ok := params["countryCode"].(string)
 	if !ok {
 		countryCode = "US"
 	}
-
-	// Libphonenumber is broken by design: Parse will try to extract the number from any text.
-	if phonenumbers.VALID_PHONE_NUMBER_PATTERN.MatchString(cred) {
-		if num, err := phonenumbers.Parse(cred, countryCode); err == nil {
-			// It's a phone number. Not checking the length because phone numbers cannot be that long.
-			if phonenumbers.IsValidNumber(num) {
-				return validatorName + ":" + phonenumbers.Format(num, phonenumbers.E164), nil
-			}
-		}
+	number, err := phonenumbers.Parse(cred, countryCode)
+	if err != nil {
+		return "", t.ErrMalformed
 	}
-	return "", t.ErrMalformed
+	if !phonenumbers.IsValidNumber(number) {
+		return "", t.ErrMalformed
+	}
+	if numType := phonenumbers.GetNumberType(number); numType != phonenumbers.FIXED_LINE_OR_MOBILE &&
+		numType != phonenumbers.MOBILE {
+		return "", t.ErrMalformed
+	}
+	return validatorName + ":" + phonenumbers.Format(number, phonenumbers.E164), nil
 }
 
-// Request sends a request for confirmation to the user: makes a record in DB  and nothing else.
-func (*validator) Request(user t.Uid, cred, lang, resp string, tmpToken []byte) (bool, error) {
-	// TODO: actually send a validation SMS or make a call to the provided `cred` here.
-	return true, nil
+// Request sends a request for confirmation to the user: makes a record in DB and nothing else.
+func (v *validator) Request(user t.Uid, phone, lang, resp string, tmpToken []byte) (bool, error) {
+	// Phone validator cannot accept an immediate response.
+	if resp != "" {
+		return false, t.ErrFailed
+	}
+
+	// Generate expected response as a random numeric string between 0 and 999999.
+	code, err := rand.Int(rand.Reader, v.maxCodeValue)
+	if err != nil {
+		return false, err
+	}
+	resp = strconv.FormatInt(code.Int64(), 10)
+	resp = strings.Repeat("0", v.CodeLength-len(resp)) + resp
+
+	var template *textt.Template
+	if v.langMatcher != nil {
+		_, idx := i18n.MatchStrings(v.langMatcher, lang)
```

Comment:
What if `lang` doesn't match?
This may lead to a crash. We need a fallback mechanism to the default template.

Reply:
Then the default language is used. That's the first language in the list of languages passed to the `NewMatcher`. That's `en` in our config and the template for English is provided.